File: src/core/typechecker/typechecker.go
package typechecker

import (
	"fmt"
	"quill/src/core/ast"
)

// ─── Type Checker ─────────────────────────────────────────────────────────────

type TypeChecker struct {
	env                   []map[string]VariableInfo
	functions             map[string]FunctionSignature
	errors                []string
	currentFuncReturnType string
}

type VariableInfo struct {
	Type     string
	Mutable  bool
	Declared bool
}

type FunctionSignature struct {
	ParamTypes []string
	ReturnType string
}

func NewTypeChecker() *TypeChecker {
	tc := &TypeChecker{
		env:       []map[string]VariableInfo{{}},
		functions: make(map[string]FunctionSignature),
	}
	return tc
}

// ─── Scope Management ─────────────────────────────────────────────────────────

func (tc *TypeChecker) pushScope() {
	tc.env = append(tc.env, make(map[string]VariableInfo))
}

func (tc *TypeChecker) popScope() {
	if len(tc.env) > 1 {
		tc.env = tc.env[:len(tc.env)-1]
	}
}

func (tc *TypeChecker) declare(name string, info VariableInfo) {
	tc.env[len(tc.env)-1][name] = info
}

func (tc *TypeChecker) lookup(name string) (VariableInfo, bool) {
	for i := len(tc.env) - 1; i >= 0; i-- {
		if info, ok := tc.env[i][name]; ok {
			return info, true
		}
	}
	return VariableInfo{}, false
}

// ─── Public API ───────────────────────────────────────────────────────────────

func (tc *TypeChecker) Check(program ast.Program) error {
	// FIXED: Reset currentFuncReturnType before checking
	tc.currentFuncReturnType = ""
	for _, node := range program.Body {
		tc.checkNode(node)
	}
	if len(tc.errors) > 0 {
		result := ""
		for _, e := range tc.errors {
			result += e + "\n"
		}
		return fmt.Errorf("%s", result)
	}
	return nil
}

// ─── Core Type Checking ───────────────────────────────────────────────────────

func (tc *TypeChecker) checkNode(node ast.Node) string {
	switch n := node.(type) {

	case ast.Program:
		for _, stmt := range n.Body {
			tc.checkNode(stmt)
		}
		return "void"

	case ast.VariableDeclaration:
		valueType := tc.checkNode(n.Value)

		if n.DeclaredType != "" && n.DeclaredType != valueType {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot assign %s to %s — '%s'",
					valueType, n.DeclaredType, n.Identifier))
		}

		finalType := n.DeclaredType
		if finalType == "" {
			finalType = valueType
		}

		if _, exists := tc.env[len(tc.env)-1][n.Identifier]; exists {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: '%s' is already declared in this scope", n.Identifier))
		}

		tc.declare(n.Identifier, VariableInfo{
			Type:     finalType,
			Mutable:  n.Mutable,
			Declared: true,
		})
		return finalType

	case ast.Assignment:
		info, exists := tc.lookup(n.Identifier)
		if !exists {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: '%s' is not declared", n.Identifier))
			return "unknown"
		}
		if !info.Mutable {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot assign to constant '%s'", n.Identifier))
		}
		valueType := tc.checkNode(n.Value)
		if info.Type != "" && info.Type != valueType {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot assign %s to %s — '%s'",
					valueType, info.Type, n.Identifier))
		}
		return valueType

	case ast.Function:
		sig := FunctionSignature{ReturnType: n.ReturnType}
		for _, p := range n.Params {
			sig.ParamTypes = append(sig.ParamTypes, p.Type)
		}
		tc.functions[n.Name] = sig

		tc.pushScope()
		savedReturnType := tc.currentFuncReturnType
		tc.currentFuncReturnType = n.ReturnType

		for _, p := range n.Params {
			tc.declare(p.Name, VariableInfo{
				Type:     p.Type,
				Mutable:  true,
				Declared: true,
			})
		}

		for _, stmt := range n.Body {
			tc.checkNode(stmt)
		}

		tc.currentFuncReturnType = savedReturnType
		tc.popScope()
		return "void"

	case ast.ReturnStatement:
		if n.Value == nil {
			if tc.currentFuncReturnType != "" && tc.currentFuncReturnType != "void" {
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: expected return value of type %s, got void",
						tc.currentFuncReturnType))
			}
			return "void"
		}
		valueType := tc.checkNode(n.Value)
		if tc.currentFuncReturnType != "" && tc.currentFuncReturnType != valueType {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: expected return type %s, got %s",
					tc.currentFuncReturnType, valueType))
		}
		return valueType

	case ast.PrintStatement:
		tc.checkNode(n.Expression)
		return "void"

	case ast.IfStatement:
		condType := tc.checkNode(n.Condition)
		if condType != "bool" && condType != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: condition must be bool, got %s", condType))
		}

		tc.pushScope()
		for _, stmt := range n.Consequent {
			tc.checkNode(stmt)
		}
		tc.popScope()

		if len(n.Alternate) > 0 {
			tc.pushScope()
			for _, stmt := range n.Alternate {
				tc.checkNode(stmt)
			}
			tc.popScope()
		}
		return "void"

	case ast.WhileLoop:
		condType := tc.checkNode(n.Condition)
		if condType != "bool" && condType != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: loop condition must be bool, got %s", condType))
		}

		tc.pushScope()
		for _, stmt := range n.Body {
			tc.checkNode(stmt)
		}
		tc.popScope()
		return "void"

	case ast.BinaryExpression:
		left := tc.checkNode(n.Left)
		right := tc.checkNode(n.Right)

		switch n.Operator {
		case "==", "!=", ">", "<", ">=", "<=":
			if left != right && left != "unknown" && right != "unknown" {
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: cannot compare %s and %s", left, right))
			}
			return "bool"
		case "&&", "||":
			if left != "bool" && left != "unknown" {
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: left operand of %s must be bool, got %s", n.Operator, left))
			}
			if right != "bool" && right != "unknown" {
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: right operand of %s must be bool, got %s", n.Operator, right))
			}
			return "bool"
		}

		// FIXED: Allow string concatenation with mixed types (coercion)
		if n.Operator == "+" && (left == "str" || right == "str") {
			return "str"
		}

		if left != right && left != "unknown" && right != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot use '%s' on %s and %s", n.Operator, left, right))
		}

		if left != "int" && left != "float" && left != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: operator '%s' requires numeric types, got %s", n.Operator, left))
		}
		return left

	case ast.UnaryExpression:
		operandType := tc.checkNode(n.Operand)
		switch n.Operator {
		case "-":
			if operandType != "int" && operandType != "float" && operandType != "unknown" {
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: unary '-' requires numeric type, got %s", operandType))
			}
			return operandType
		case "!":
			if operandType != "bool" && operandType != "unknown" {
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: unary '!' requires bool, got %s", operandType))
			}
			return "bool"
		}
		return "unknown"

	case ast.CallExpression:
		sig, exists := tc.functions[n.Callee]
		if !exists {
			switch n.Callee {
			case "len":
				if len(n.Arguments) != 1 {
					tc.errors = append(tc.errors, "TypeError: 'len' expects 1 argument")
				}
				return "int"
			case "toString":
				if len(n.Arguments) != 1 {
					tc.errors = append(tc.errors, "TypeError: 'toString' expects 1 argument")
				}
				return "str"
			default:
				tc.errors = append(tc.errors,
					fmt.Sprintf("TypeError: unknown function '%s'", n.Callee))
				return "unknown"
			}
		}

		if len(n.Arguments) != len(sig.ParamTypes) {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: '%s' expects %d arguments but got %d",
					n.Callee, len(sig.ParamTypes), len(n.Arguments)))
		} else {
			for i, arg := range n.Arguments {
				argType := tc.checkNode(arg)
				if sig.ParamTypes[i] != "" && argType != sig.ParamTypes[i] && argType != "unknown" {
					tc.errors = append(tc.errors,
						fmt.Sprintf("TypeError: argument %d of '%s' expects %s but got %s",
							i+1, n.Callee, sig.ParamTypes[i], argType))
				}
			}
		}
		return sig.ReturnType

	case ast.IndexExpression:
		objectType := tc.checkNode(n.Object)
		indexType := tc.checkNode(n.Index)

		if indexType != "int" && indexType != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: index must be int, got %s", indexType))
		}

		if objectType == "str" {
			return "str"
		}
		return "unknown"

	case ast.Identifier:
		info, exists := tc.lookup(n.Name)
		if !exists {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: '%s' is not declared", n.Name))
			return "unknown"
		}
		return info.Type

	case ast.Increment:
		info, exists := tc.lookup(n.Identifier)
		if !exists {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: '%s' is not declared", n.Identifier))
			return "int"
		}
		if info.Type != "int" && info.Type != "float" && info.Type != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot increment %s '%s'", info.Type, n.Identifier))
		}
		if !info.Mutable {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot increment constant '%s'", n.Identifier))
		}
		return info.Type

	case ast.Decrement:
		info, exists := tc.lookup(n.Identifier)
		if !exists {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: '%s' is not declared", n.Identifier))
			return "int"
		}
		if info.Type != "int" && info.Type != "float" && info.Type != "unknown" {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot decrement %s '%s'", info.Type, n.Identifier))
		}
		if !info.Mutable {
			tc.errors = append(tc.errors,
				fmt.Sprintf("TypeError: cannot decrement constant '%s'", n.Identifier))
		}
		return info.Type

	case ast.ExpressionStatement:
		return tc.checkNode(n.Expression)

	case ast.LiteralInt:
		return "int"
	case ast.LiteralFloat:
		return "float"
	case ast.LiteralString:
		return "str"
	case ast.LiteralBool:
		return "bool"
	case ast.Comment:
		return "void"
	}

	tc.errors = append(tc.errors,
		fmt.Sprintf("TypeError: unknown AST node type %T", node))
	return "unknown"
}

File: src/core/typechecker/typechecker.cpp
#pragma once

#include <stdexcept>
#include <string>
#include <map>
#include <unordered_map>
#include <utility>
#include <vector>

#include "../ast/ast.cpp"

class TypeChecker {
public:
    // functionReturnTypes is the transpiler's own already-resolved map
    // of function name -> return type (explicit annotation, or
    // inferred from the function's own return statements when one
    // wasn't given) -- see buildFunctionReturnTypes() in trans.cpp.
    // Reusing it here means this checker and the code actually emitted
    // by the transpiler can never disagree about what a function
    // returns.
    void check(const Program& program, const std::map<std::string, std::string>& functionReturnTypes) {
        this->functionReturnTypes = functionReturnTypes;
        for (Node* stmt : program.body) {
            checkNode(stmt);
        }
    }

private:
    std::unordered_map<std::string, std::string> symbols;
    std::map<std::string, std::string> functionReturnTypes;

    void checkNode(Node* node) {
        if (!node) return;

        if (auto* decl = dynamic_cast<VariableDeclaration*>(node)) {
            checkVariableDeclaration(decl);
            return;
        }

        if (auto* assign = dynamic_cast<Assignment*>(node)) {
            checkAssignment(assign);
            return;
        }

        if (auto* expr = dynamic_cast<ExpressionStatement*>(node)) {
            checkNode(expr->expression);
            return;
        }

        if (auto* fn = dynamic_cast<Function*>(node)) {
            auto saved = symbols;
            for (const Param& param : fn->params) {
                symbols[param.name] = param.type;
            }

            for (Node* stmt : fn->body) {
                checkNode(stmt);
            }

            symbols = saved;
            return;
        }

        if (auto* ifStmt = dynamic_cast<IfStatement*>(node)) {
            checkNode(ifStmt->condition);
            for (Node* stmt : ifStmt->consequent) checkNode(stmt);
            for (Node* stmt : ifStmt->alternate) checkNode(stmt);
            return;
        }

        if (auto* whileLoop = dynamic_cast<WhileLoop*>(node)) {
            checkNode(whileLoop->condition);
            for (Node* stmt : whileLoop->body) checkNode(stmt);
            return;
        }

        if (auto* ret = dynamic_cast<ReturnStatement*>(node)) {
            checkNode(ret->value);
            return;
        }

        if (auto* print = dynamic_cast<PrintStatement*>(node)) {
            checkNode(print->expression);
            return;
        }

        if (auto* call = dynamic_cast<CallExpression*>(node)) {
            for (Node* arg : call->arguments) {
                checkNode(arg);
            }
            return;
        }

        if (auto* bin = dynamic_cast<BinaryExpression*>(node)) {
            checkNode(bin->left);
            checkNode(bin->right);
            return;
        }

        if (auto* unary = dynamic_cast<UnaryExpression*>(node)) {
            checkNode(unary->operand);
            return;
        }

        if (auto* idx = dynamic_cast<IndexExpression*>(node)) {
            checkNode(idx->object);
            checkNode(idx->index);
            std::string objType = inferType(idx->object);
            bool isArray = objType.find('[') != std::string::npos;
            if (objType != "string" && !isArray) {
                throw std::runtime_error("indexing with [] is only supported on string or array values");
            }
            return;
        }

        if (auto* idxAssign = dynamic_cast<IndexAssignment*>(node)) {
            checkNode(idxAssign->object);
            checkNode(idxAssign->index);
            checkNode(idxAssign->value);

            std::string objType = inferType(idxAssign->object);
            auto bracketPos = objType.find('[');
            if (bracketPos == std::string::npos) {
                throw std::runtime_error("assignment with [] is only supported on array values");
            }

            std::string elemType = objType.substr(0, bracketPos);
            std::string valueType = inferType(idxAssign->value);
            if (valueType != elemType) {
                throw std::runtime_error("type mismatch assigning into array");
            }
            return;
        }


        if (auto* lit = dynamic_cast<ArrayLiteral*>(node)) {
            for (Node* el : lit->elements) {
                checkNode(el);
            }
            return;
        }

        if (auto* ident = dynamic_cast<Identifier*>(node)) {
            if (symbols.find(ident->name) == symbols.end()) {
                throw std::runtime_error("undefined identifier: " + ident->name);
            }
            return;
        }
    }

    // C_call()/C_top() splice raw, unverifiable C text -- Quill has no
    // way to know their real type, so (the same trust model as `extern
    // func`) the student's own annotation is trusted rather than
    // checked against a guess that can only ever be wrong.
    bool isRawCSplice(Node* node) {
        auto* call = dynamic_cast<CallExpression*>(node);
        return call && (call->callee == "C_call" || call->callee == "C_top");
    }

    void checkVariableDeclaration(VariableDeclaration* decl) {
        if (!decl->value) {
            // Arrays (and any other declaration without an initializer)
            // still need to be registered so later uses aren't flagged
            // as undefined identifiers.
            if (!decl->declaredType.empty()) {
                symbols[decl->identifier] = decl->declaredType;
            }
            return;
        }
        checkNode(decl->value);

        if (isRawCSplice(decl->value)) {
            symbols[decl->identifier] = decl->declaredType.empty() ? "unknown" : decl->declaredType;
            return;
        }

        std::string valueType = inferType(decl->value);
        if (!decl->declaredType.empty()) {
            // Fixed number[3] may be initialized from number[] literal of length 3
            // Dynamic number[] matches number[] literal
            bool declArr = decl->declaredType.find('[') != std::string::npos;
            bool valArr = valueType.find('[') != std::string::npos;
            if (declArr && valArr) {
                std::string declElem = decl->declaredType.substr(0, decl->declaredType.find('['));
                std::string valElem = valueType.substr(0, valueType.find('['));
                if (declElem != valElem) {
                    throw std::runtime_error("type mismatch for variable: " + decl->identifier);
                }
                // length check for fixed-size left to runtime / optional later
            } else if (decl->declaredType != valueType) {
                throw std::runtime_error("type mismatch for variable: " + decl->identifier);
            }
            symbols[decl->identifier] = decl->declaredType;
        } else {
            symbols[decl->identifier] = valueType;
        }
    }

    void checkAssignment(Assignment* assign) {
        auto it = symbols.find(assign->identifier);
        if (it == symbols.end()) {
            throw std::runtime_error("assignment to undefined variable: " + assign->identifier);
        }

        if (!assign->value) return;
        checkNode(assign->value);

        if (isRawCSplice(assign->value)) {
            return;
        }

        std::string valueType = inferType(assign->value);
        if (valueType != it->second) {
            throw std::runtime_error("type mismatch on assignment to: " + assign->identifier);
        }
    }

    std::string inferType(Node* node) {
        if (!node) return "void";

        if (dynamic_cast<LiteralInt*>(node)) return "number";
        if (dynamic_cast<LiteralDoudle*>(node)) return "double";
        if (dynamic_cast<LiteralString*>(node)) return "string";
        if (dynamic_cast<LiteralBool*>(node)) return "bool";

        if (auto* ident = dynamic_cast<Identifier*>(node)) {
            auto it = symbols.find(ident->name);
            if (it != symbols.end()) return it->second;
            return "unknown";
        }

        // Indexing: string -> char code (number); T[N]/T[] -> T
        if (auto* idx = dynamic_cast<IndexExpression*>(node)) {
            std::string objType = inferType(idx->object);
            auto bracketPos = objType.find('[');
            if (bracketPos != std::string::npos)
                return objType.substr(0, bracketPos);
            if (objType == "string")
                return "number";
            return "number";
        }

        if (auto* lit = dynamic_cast<ArrayLiteral*>(node)) {
            if (lit->elements.empty())
                return "number[]"; // default empty literal
            std::string elem = inferType(lit->elements[0]);
            for (size_t i = 1; i < lit->elements.size(); i++) {
                std::string t = inferType(lit->elements[i]);
                if (t != elem) {
                    // allow number/double promotion inside literals
                    if ((elem == "number" && t == "double") || (elem == "double" && t == "number"))
                        elem = "double";
                    else
                        return "unknown";
                }
            }
            return elem + "[]";
        }

        if (auto* binary = dynamic_cast<BinaryExpression*>(node)) {
            std::string left = inferType(binary->left);
            std::string right = inferType(binary->right);
            if (left == right) return left;
            if ((left == "number" && right == "double") || (left == "double" && right == "number")) return "double";
            return "unknown";
        }

        if (auto* unary = dynamic_cast<UnaryExpression*>(node)) {
            return inferType(unary->operand);
        }

        if (auto* call = dynamic_cast<CallExpression*>(node)) {
            auto it = functionReturnTypes.find(call->callee);
            return it != functionReturnTypes.end() ? it->second : "unknown";
        }
        if (dynamic_cast<Assignment*>(node)) return "void";

        return "unknown";
    }
};